File: KantaroBLL/Accordo.vb
Public Class Accordo
    Implements ICloneable

    Private _scala As Scala
    Public Property Basso As Nota
    Private _estensione As EstensioneT
    'Private _formaPreferita As FormaPreferita

    Public Sub New(scala As Scala)
        Me.New(scala, EstensioneT.Empty, Nothing)
    End Sub

    Private Sub New(scala As Scala, testoPreferito As String)
        Me.New(scala)
        '_formaPreferita = New FormaPreferita With {.Scala = scala, .Testo = testoPreferito}
    End Sub

    Public Sub New(scala As Scala, estensione As EstensioneT)
        Me.New(scala, estensione, Nothing)
    End Sub

    Public Sub New(scala As Scala, num As Integer)
        Me.New(scala, num, EstensioneT.EstensioneVariazioneSemitonoEnum.None, Nothing)
    End Sub

    Public Sub New(scala As Scala, num As Integer, varEstensione As EstensioneT.EstensioneVariazioneSemitonoEnum)
        Me.New(scala, num, varEstensione, Nothing)
    End Sub

    Public Sub New(scala As Scala, num As Integer, varEstensione As EstensioneT.EstensioneVariazioneSemitonoEnum, basso As Nota)
        _scala = scala
        If num > 0 Then
            Dim ext As New EstensioneT
            ext.Valore = num
            ext.VariazioneSemitono = varEstensione
            Me.Estensione = ext
        End If
        Me.Basso = basso
    End Sub

    Public Sub New(scala As Scala, estensione As EstensioneT, basso As Nota)
        _scala = scala
        Me.Estensione = estensione
        Me.Basso = basso
    End Sub

    Public ReadOnly Property Scala As Scala
        Get
            Return _scala
        End Get
    End Property

    Private Structure FormaPreferita
        Public Property Scala As Scala
        Public Property Testo As String
    End Structure

    Public Class EstensioneT
        Public Enum EstensioneVariazioneSemitonoEnum
            None = 0
            Diminuito = 1
            Aumentato = 2
        End Enum

        Public Property Valore As Integer
        Public Property VariazioneSemitono As EstensioneVariazioneSemitonoEnum

        Public Shared ReadOnly Property Empty As EstensioneT
            Get
                Return New EstensioneT()
            End Get
        End Property

        Public ReadOnly Property IsEmpty As Boolean
            Get
                Return Me.Valore = 0
            End Get
        End Property

        Public Overrides Function Equals(obj As Object) As Boolean
            If obj Is Nothing Then Return False
            If Not TypeOf (obj) Is EstensioneT Then Return False
            Dim o = DirectCast(obj, EstensioneT)
            Return o.Valore = Me.Valore AndAlso o.VariazioneSemitono = Me.VariazioneSemitono
        End Function

        Public Overrides Function ToString() As String
            Dim sb As New System.Text.StringBuilder
            If Valore > 0 Then sb.Append(Valore)
            If Me.VariazioneSemitono = EstensioneVariazioneSemitonoEnum.Aumentato Then sb.Append("+")
            If Me.VariazioneSemitono = EstensioneVariazioneSemitonoEnum.Diminuito Then sb.Append("dim")
            Return sb.ToString()
        End Function

        Public Shared Widening Operator CType(value As Integer) As EstensioneT
            Dim e As New EstensioneT
            e.Valore = value
            Return e
        End Operator

        Public Shared Function TryParse(value As String, ByRef result As EstensioneT) As Boolean
            If value Is Nothing Then Throw New ArgumentNullException("value")
            Dim r = New EstensioneT
            If Not String.IsNullOrWhiteSpace(value) Then
                Dim regEst As New System.Text.RegularExpressions.Regex("^[0-9]+")
                If Not regEst.IsMatch(value) Then Return False
                r.Valore = CInt(regEst.Match(value).Value)
                Select Case value.Remove(0, regEst.Match(value).Value.Length)
                    Case "+"
                        r.VariazioneSemitono = EstensioneVariazioneSemitonoEnum.Aumentato
                    Case "dim"
                        r.VariazioneSemitono = EstensioneVariazioneSemitonoEnum.Diminuito
                    Case Else
                        r.VariazioneSemitono = EstensioneVariazioneSemitonoEnum.None
                End Select
            End If
            result = r
            Return True
        End Function

        Friend Sub New()

        End Sub

        Public Sub New(valore As Integer, varSemitono As EstensioneVariazioneSemitonoEnum)
            Me.Valore = valore
            Me.VariazioneSemitono = varSemitono
        End Sub
    End Class

    Public Property Estensione As EstensioneT
        Get
            Return If(_estensione Is Nothing OrElse _estensione.IsEmpty, Nothing, _estensione)
        End Get
        Set(value As EstensioneT)
            _estensione = value
        End Set
    End Property

#Region " Definizione degli Operatori "
    Public Shared Operator +(ByVal accordo1 As Accordo, ByVal distanza As Distanza) As Accordo
        If distanza.Semitoni = 0 Then Return accordo1
        Dim n = New Accordo(New Scala(accordo1.Scala.NotaFondamentale + distanza, accordo1.Scala.Modo), accordo1.Estensione)
        If accordo1.Basso IsNot Nothing Then n.Basso = accordo1.Basso + distanza

        Return n
    End Operator
    Public Shared Operator -(ByVal accordo1 As Accordo, ByVal distanza As Distanza) As Accordo
        If distanza.Semitoni = 0 Then Return accordo1
        Dim n = New Accordo(New Scala(accordo1.Scala.NotaFondamentale - distanza, accordo1.Scala.Modo), accordo1.Estensione)
        If accordo1.Basso IsNot Nothing Then n.Basso = accordo1.Basso - distanza

        Return n
    End Operator

    Public Shared Operator =(ByVal class1 As Accordo, ByVal class2 As Accordo) As Boolean
        If class1 Is Nothing Then Return False
        Return class1.Equals(class2)
    End Operator

    Public Shared Operator <>(ByVal class1 As Accordo, ByVal class2 As Accordo) As Boolean
        If class1 Is Nothing Then Return False
        Return Not class1.Equals(class2)
    End Operator
#End Region

    Public Overrides Function Equals(obj As Object) As Boolean
        If Not TypeOf obj Is Accordo Then Return False
        Dim a = DirectCast(obj, Accordo)
        Dim bScala = a.Scala.Equals(Scala)
        Dim bBasso = ((Basso Is Nothing AndAlso a.Basso Is Nothing) OrElse
                      (Basso IsNot Nothing AndAlso a.Basso IsNot Nothing AndAlso Basso.Equals(a.Basso)))
        Dim bAlter = ((Estensione Is Nothing AndAlso a.Estensione Is Nothing) OrElse
                      (Estensione IsNot Nothing AndAlso a.Estensione IsNot Nothing AndAlso Estensione.Equals(a.Estensione)))
        Return bScala AndAlso bAlter AndAlso bBasso AndAlso bAlter
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If obj1 Is Nothing AndAlso obj2 Is Nothing Then Return True
        If Not TypeOf obj1 Is Accordo Then Return False
        Return DirectCast(obj1, Accordo).Equals(obj2)
    End Function

    Public Shared Widening Operator CType(value As String) As Accordo
        Dim a As Accordo = Nothing
        TryParse(value, a)
        Return a
    End Operator

    Public Shared Function TryParse(value As String, ByRef result As Accordo) As Boolean
        If value Is Nothing Then Throw New ArgumentNullException("text")
        Dim parti = value.Split("/")
        Dim sc As Scala = Nothing
        If Not Scala.TryParse(parti(0), sc) Then Throw New InvalidCastException(String.Format("Testo non convertibile in Accordo: {0}", parti(0)))

        Dim acc As New Accordo(sc, value)
        If parti(0).Remove(0, sc.ToString.Length).Length > 0 Then
            Dim temp = parti(0).Remove(0, sc.ToString.Length)
            Dim ext As New EstensioneT
            Dim num = New String(temp.ToCharArray.TakeWhile(Function(c As Char) Char.IsDigit(c)).ToArray)
            Dim valNumb As Integer
            If Integer.TryParse(num, valNumb) Then ext.Valore = valNumb
            If num <> temp Then
                Select Case temp.Remove(0, num.Length)
                    Case "+" : ext.VariazioneSemitono = EstensioneT.EstensioneVariazioneSemitonoEnum.Aumentato
                    Case "dim" : ext.VariazioneSemitono = EstensioneT.EstensioneVariazioneSemitonoEnum.Diminuito
                    Case Else : ext.VariazioneSemitono = EstensioneT.EstensioneVariazioneSemitonoEnum.None
                End Select
            End If
            acc.Estensione = ext
        End If
        If parti.Length > 1 Then
            Dim basso As Nota = Nothing
            If parti.Length <> 2 OrElse Not Nota.TryParse(parti(1), basso) Then Throw New InvalidCastException(String.Format("Testo non convertibile in Accordo: {0}", parti(1)))
            acc.Basso = basso
        End If
        result = acc
        Return True
    End Function

    Public Overrides Function ToString() As String
        Return Me.ToString(False)
    End Function

    Public Overloads Function ToString(bemollePrefer As Boolean) As String
        Dim sb As New System.Text.StringBuilder
        'If _formaPreferita.Scala IsNot Nothing AndAlso Not String.IsNullOrWhiteSpace(_formaPreferita.Testo) AndAlso Scala.Equals(_formaPreferita.Scala) Then
        '    sb.Append(_formaPreferita.Testo)
        'Else
        sb.Append(Me.Scala.ToString(bemollePrefer))
        If TypeOf Scala.Modo Is ModoMinoreArmonica Then
            sb.Append("-")
        End If
        If Estensione IsNot Nothing Then
            sb.Append(Estensione.ToString)
        End If
        If Me.Basso IsNot Nothing Then
            sb.Append("/")
            sb.Append(Me.Basso.ToString)
        End If
        'End If
        Return sb.ToString()
    End Function

    Public Function Clone() As Object Implements ICloneable.Clone
        Dim n As New Accordo(Scala, Me.Estensione, Me.Basso)
        'n._formaPreferita = _formaPreferita
        Return n
    End Function
End Class

File: Enzo.Music/Accordo.cs
using static System.Net.Mime.MediaTypeNames;
using static System.Runtime.InteropServices.JavaScript.JSType;
using System.Security.Claims;
using System.Text.RegularExpressions;
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Runtime.InteropServices;

namespace Enzo.Music;

public class Accordo : ICloneable, IAdditionOperators<Accordo, Distanza, Accordo>, ISubtractionOperators<Accordo, Distanza, Accordo>
    , IEqualityOperators<Accordo, Accordo, bool>
{


    private struct FormaPreferita(Scala scala, string testo)
    {
        public Scala Scala = scala;
        public string Testo = testo;
    }

    public class EstensioneT
    {
        public enum EstensioneVariazioneSemitonoEnum
        {
            None = 0,
            Diminuito = 1,
            Aumentato = 2
        }

        public int Valore;
        public EstensioneVariazioneSemitonoEnum VariazioneSemitono;

        public static EstensioneT Empty => new();

        public bool IsEmpty => Valore == 0;

        public override bool Equals(object? obj)
        {
            if (obj is null || obj is not EstensioneT) return false;
            EstensioneT o = obj as EstensioneT ?? Empty;
            return Valore.Equals(o.Valore) && VariazioneSemitono.Equals(o.VariazioneSemitono);
        }
        public override string ToString()
        {
            System.Text.StringBuilder sb = new();
            if (Valore > 0) sb.Append(Valore);
            if (VariazioneSemitono == EstensioneVariazioneSemitonoEnum.Aumentato) sb.Append('+');
            if (VariazioneSemitono == EstensioneVariazioneSemitonoEnum.Diminuito) sb.Append("dim");
            return sb.ToString();
        }

        public override int GetHashCode()
        {
            return ToString().GetHashCode();
        }

        //public static explicit operator EstensioneT(int value)
        //{
        //    return new EstensioneT { Valore = value };
        //}

        public static implicit operator EstensioneT(int value)
        {
            TryParse(value, out EstensioneT ext);
            return ext;
        }

        public static bool TryParse(string value, out EstensioneT? result)
        {
            ArgumentNullException.ThrowIfNull(value);
            EstensioneT r = new();
            if (!string.IsNullOrWhiteSpace(value))
            {
                var regEst = new Regex("^[0-9]+");
                if (!regEst.IsMatch(value))
                {
                    result = null;
                    return false;
                }

                r.Valore = int.Parse(regEst.Match(value).Value);
                r.VariazioneSemitono = value[regEst.Match(value).Value.Length..] switch
                {
                    "+" => EstensioneVariazioneSemitonoEnum.Aumentato,
                    "dim" => EstensioneVariazioneSemitonoEnum.Diminuito,
                    _ => EstensioneVariazioneSemitonoEnum.None,
                };
            }
            result = r;
            return true;
        }

        public static bool TryParse(int value, out EstensioneT result)
        {
            EstensioneT r = new()
            {
                Valore = value,
                VariazioneSemitono = EstensioneVariazioneSemitonoEnum.None
            };
            result = r;
            return true;
        }

        internal EstensioneT() { }

        public EstensioneT(int valore, EstensioneVariazioneSemitonoEnum varSemitono)
        {
            Valore = valore;
            VariazioneSemitono = varSemitono;
        }
    }

    private readonly Scala scala;

    public Accordo(Scala scala) : this(scala, EstensioneT.Empty)
    {
    }

    private Accordo(Scala scala, string testoPreferito) : this(scala)
    {
    }

    public Accordo(Scala scala, EstensioneT estensione)
    {
        this.scala = scala;
        this.Estensione = estensione;
    }

    public Accordo(Scala scala, int num) : this(scala, num, EstensioneT.EstensioneVariazioneSemitonoEnum.None)
    {
    }

    public Accordo(Scala scala, int num, EstensioneT.EstensioneVariazioneSemitonoEnum varEstensione)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(num);
        this.scala = scala;
        Estensione = new() { Valore = num, VariazioneSemitono = varEstensione };
    }

    public Accordo(Scala scala, int num, EstensioneT.EstensioneVariazioneSemitonoEnum varEstensione, Nota basso) : this(scala, num, varEstensione)
    {
        Basso = basso;
    }
    public Accordo(Scala scala, EstensioneT estensione, Nota basso) : this(scala, estensione)
    {
        Basso = basso;
    }

    public Scala Scala => scala;
    public Nota? Basso { get; set; }
    public EstensioneT Estensione { get; set; }

    #region Definizione degli Operatori
    public static Accordo operator +(Accordo left, Distanza right)
    {
        if (right.Semitoni == 0) return left;
        Accordo n = new(new Scala(left.Scala.NotaFondamentale + right, left.Scala.Modo), left.Estensione);
        if (left.Basso is not null) n.Basso = left.Basso + right;

        return n;
    }

    public static Accordo operator -(Accordo left, Distanza right)
    {
        if (right.Semitoni == 0) return left;
        Accordo n = new(new Scala(left.Scala.NotaFondamentale - right, left.Scala.Modo), left.Estensione);
        if (left.Basso is not null) n.Basso = left.Basso - right;

        return n;
    }

    public static bool operator ==(Accordo? left, Accordo? right)
    {
        if (left is null || right is null) return false;
        return left.Equals(right);
    }

    public static bool operator !=(Accordo? left, Accordo? right)
    {
        if (left is null || right is null) return false;
        return !left.Equals(right);
    }

    public override bool Equals(object? obj)
    {
        if (obj is null || obj is not Accordo) return false;
        Accordo a = (Accordo)obj;
        bool bScala = a.Scala.Equals(Scala);
        bool bBasso = Basso is null && a.Basso is null ||
                      Basso is not null && a.Basso is not null && Basso.Equals(a.Basso);
        bool bAlter = Estensione is null && a.Estensione is null ||
                      Estensione is not null && a.Estensione is not null && Estensione.Equals(a.Estensione);
        return bScala && bAlter && bBasso && bAlter;
    }

    public static bool Equals(Accordo? left, Accordo? right)
    {
        if (left is null || right is null) return false;
        return left.Equals(right);
    }

    #endregion

    public static bool TryParse(string value, out Accordo? result)
    {
        ArgumentNullException.ThrowIfNull(value);
        var parti = value.Split("/");
        if (!Scala.TryParse(parti[0], out Scala? sc)) throw new InvalidCastException($"Testo non convertibile in Accordo: {parti[0]}");
        if (sc is null) throw new InvalidCastException();
        Accordo acc = new(sc, value);
        if (parti[0][sc.ToString().Length..].Length > 0)
        {
            string temp = parti[0][sc.ToString().Length..];
            EstensioneT ext = new();
            string num = new([.. temp.ToCharArray().TakeWhile(char.IsDigit)]);
            if (int.TryParse(num, out int valNumb)) ext.Valore = valNumb;
            if (num != temp)
            {
                ext.VariazioneSemitono = temp[num.Length..] switch
                {
                    "+" => EstensioneT.EstensioneVariazioneSemitonoEnum.Aumentato,
                    "dim" => EstensioneT.EstensioneVariazioneSemitonoEnum.Diminuito,
                    _ => EstensioneT.EstensioneVariazioneSemitonoEnum.None,
                };
            }
            acc.Estensione = ext;
        }
        if (parti.Length > 1)
        {
            if (parti.Length != 2 || !Nota.TryParse(parti[1], out Nota? basso)) throw new InvalidCastException($"Testo non convertibile in Accordo: {parti[1]}");
            acc.Basso = basso;
        }
        result = acc;
        return true;
    }

    public static explicit operator Accordo?(string value)
    {
        _ = TryParse(value, out Accordo? a);
        return a;
    }

    public override string ToString()
    {
        return ToString(false);
    }
    public string ToString(bool bemollePrefer)
    {
        System.Text.StringBuilder sb = new();
        sb.Append(Scala.ToString(bemollePrefer));
        if (Scala.Modo is ModoMinoreArmonica) sb.Append('-');
        if (Estensione is not null) sb.Append(Estensione.ToString());
        if (Basso is not null)
        {
            sb.Append('/');
            sb.Append(Basso.ToString());
        }
        return sb.ToString();
    }

    public override int GetHashCode()
    {
        return ToString().GetHashCode();
    }

    public object Clone()
    {
        Accordo n = Basso is null ? new(scala, Estensione) : new(scala, Estensione, Basso);
        return n;
    }
}
